firestore/bulkwriter.go

```diff
@@ -0,0 +1,332 @@
+	return BulkWriterStatus{
+		IsOpen: bw.isOpen,
+	}
+}
+
+// Create adds a document creation write to the queue of writes to send.
+func (bw *BulkWriter) Create(doc *DocumentRef, datum interface{}) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newCreateWrites(datum)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot create %v with %v", doc.ID, datum)
+	}
+
+	return &BulkWriterJob{
+		op: bw.write(w[0]),
+	}, nil
+}
+
+// Delete adds a document deletion write to the queue of writes to send.
+func (bw *BulkWriter) Delete(doc *DocumentRef, preconds ...Precondition) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newDeleteWrites(preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot delete doc %v", doc.ID)
+	}
+
+	return &BulkWriterJob{
+		op: bw.write(w[0]),
+	}, nil
+}
+
+// Set adds a document set write to the queue of writes to send.
+func (bw *BulkWriter) Set(doc *DocumentRef, datum interface{}, opts ...SetOption) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newSetWrites(datum, opts)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot set %v on doc %v", datum, doc.ID)
+	}
+
+	return &BulkWriterJob{
+		op: bw.write(w[0]),
+	}, nil
+}
+
+// Update adds a document update write to the queue of writes to send.
+func (bw *BulkWriter) Update(doc *DocumentRef, updates []Update, preconds ...Precondition) (*BulkWriterJob, error) {
+	err := bw.checkWriteConditions(doc)
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := doc.newUpdatePathWrites(updates, preconds)
+	if err != nil {
+		return nil, fmt.Errorf("firestore: cannot update doc %v", doc.ID)
+	}
+	return &BulkWriterJob{
+		op: bw.write(w[0]),
+	}, nil
+}
+
+// checkConditions determines whether this write attempt is valid. It returns
+// an error if either the BulkWriter has already been closed or if it
+// receives a nil document reference.
+func (bw *BulkWriter) checkWriteConditions(doc *DocumentRef) error {
+	if !bw.isOpen {
+		return errors.New("firestore: BulkWriter has been closed")
+	}
+
+	if doc == nil {
+		return errors.New("firestore: nil document contents")
+	}
+
+	_, havePath := bw.docUpdatePaths[doc.shortPath]
+	if havePath {
+		return fmt.Errorf("firestore: bulkwriter: received duplicate write for path: %v", doc.shortPath)
+	}
+
+	bw.docUpdatePaths[doc.shortPath] = true
+
+	return nil
+}
+
+// write packages up write requests into bulkWriterJob objects.
+func (bw *BulkWriter) write(w *pb.Write) bulkWriterOperation {
+	r := make(chan *pb.WriteResult, 1)
+	e := make(chan error, 1)
+
+	j := bulkWriterOperation{
+		result: r,
+		write:  w,
+		err:    e,
+		size:   0, // ignore operation size constraints; can't be inferred at compile time
+	}
+
+	err := bw.bundler.Add(j, j.size)
+	if err != nil {
+		j.err <- err
```

Comment:
I wonder if it is better to let users subscribe to errors with a function? I know some of our other cliets do this sort of things for async workloads.